File: src/main/java/mstparser/old/DependencyDecoder.java
package mstparser.old;

import mstparser.DependencyInstance;
import mstparser.FeatureVector;
import mstparser.KBestParseForest;
import mstparser.ParseForestItem;

import java.util.ArrayList;
import java.util.Arrays;
import gnu.trove.map.hash.TIntIntHashMap;

import scala.Tuple2;

public abstract class DependencyDecoder {
  public abstract mstparser.DependencyPipe pipe(); 
  protected abstract int[][] getTypes(double[][][][] nt_probs, int len);
  protected abstract boolean[][] calcChilds(int[] par);

    // static type for each edge: run time O(n^3 + Tn^2) T is number of types
    public abstract Tuple2<FeatureVector, String>[] decodeProjective(int len,
				       FeatureVector[][][] fvs,
				       double[][][] probs,
				       FeatureVector[][][][] nt_fvs,
				       double[][][][] nt_probs, int K);

    public Tuple2<FeatureVector, String>[] decodeNonProjective(int len,
					  FeatureVector[][][] fvs,
					  double[][][] probs,
					  FeatureVector[][][][] nt_fvs,
					  double[][][][] nt_probs, int K) {

	int[][] oldI = new int[len][len];
	int[][] oldO = new int[len][len];
	double[][] scoreMatrix = new double[len][len];
	double[][] orig_scoreMatrix = new double[len][len];
	boolean[] curr_nodes = new boolean[len];
	TIntIntHashMap[] reps = new TIntIntHashMap[len];

	int[][] static_types = null;
	if(this.pipe().getLabeled()) {
	    static_types = getTypes(nt_probs,len);
	}

	for(int i = 0; i < len; i++) {
	    curr_nodes[i] = true;
	    reps[i] = new TIntIntHashMap();
	    reps[i].put(i,0);
	    for(int j = 0; j < len; j++) {
		// score of edge (i,j) i --> j
		scoreMatrix[i][j] = probs[i < j ? i : j][i < j ? j : i][i < j ? 0 : 1]
		    + (this.pipe().getLabeled() ? nt_probs[i][static_types[i][j]][i < j ? 0 : 1][1]
		       + nt_probs[j][static_types[i][j]][i < j ? 0 : 1][0]
		       : 0.0);
		orig_scoreMatrix[i][j] = probs[i < j ? i : j][i < j ? j : i][i < j ? 0 : 1]
		    + (this.pipe().getLabeled() ? nt_probs[i][static_types[i][j]][i < j ? 0 : 1][1]
		       + nt_probs[j][static_types[i][j]][i < j ? 0 : 1][0]
		       : 0.0);
		oldI[i][j] = i;
		oldO[i][j] = j;
		if(i == j || j == 0) continue; // no self loops of i --> 0
	    }
	}

	TIntIntHashMap final_edges = chuLiuEdmonds(scoreMatrix,curr_nodes,oldI,oldO,false,new TIntIntHashMap(),reps);
	int[] par = new int[len];
	int[] ns = final_edges.keys();
	for(int i = 0; i < ns.length; i++) {
	    int ch = ns[i]; int pr = final_edges.get(ns[i]);
	    par[ch] = pr;
	}

	int[] n_par = getKChanges(par,orig_scoreMatrix,Math.min(K,par.length));
	int new_k = 1;
	for(int i = 0; i < n_par.length; i++)
	    if(n_par[i] > -1) new_k++;

	// Create Feature Vectors;
	int[][] fin_par = new int[new_k][len];
	FeatureVector[][] fin_fv = new FeatureVector[new_k][len];
	fin_par[0] = par;
	int c = 1;
	for(int i = 0; i < n_par.length; i++) {
	    if(n_par[i] > -1) {
		int[] t_par = new int[par.length];
		for(int j = 0; j < t_par.length; j++)
		    t_par[j] = par[j];
		t_par[i] = n_par[i];
		fin_par[c] = t_par;
		c++;
	    }
	}
	for(int k = 0; k < fin_par.length; k++) {
	    for(int i = 0; i < fin_par[k].length; i++) {
		int ch = i; int pr = fin_par[k][i];
		if(pr != -1) {
		    fin_fv[k][ch] = fvs[ch < pr ? ch : pr][ch < pr ? pr : ch][ch < pr ? 1 : 0];
		    if(this.pipe().getLabeled()) {
			fin_fv[k][ch] = 
			    fin_fv[k][ch].cat(nt_fvs[ch][static_types[pr][ch]][ch < pr ? 1 : 0][0]);
			fin_fv[k][ch] = 
			    fin_fv[k][ch].cat(nt_fvs[pr][static_types[pr][ch]][ch < pr ? 1 : 0][1]);
		    }
		}
		else
		    fin_fv[k][ch] = new FeatureVector();
	    }
	}
	
	
	FeatureVector[] fin = new FeatureVector[new_k];
	String[] result = new String[new_k];
	for(int k = 0; k < fin.length; k++) {
	    fin[k] = new FeatureVector();
	    for(int i = 1; i < fin_fv[k].length; i++)
		fin[k] = fin_fv[k][i].cat(fin[k]);
	    result[k] = "";
	    for(int i = 1; i < par.length; i++)
		result[k] += fin_par[k][i]+"|"+i + (this.pipe().getLabeled() ? ":"+static_types[fin_par[k][i]][i] : ":0") +" ";
	}
		
	// create d.
	Tuple2<FeatureVector, String>[] d = (Tuple2<FeatureVector, String>[]) new Tuple2[new_k];

	for(int k = 0; k < new_k; k++) {
	    d[k] = new Tuple2<FeatureVector, String>(fin[k], result[k].trim());
	}

	return d;
    }

    private int[] getKChanges(int[] par, double[][] scoreMatrix, int K) {
	int[] result = new int[par.length];
	int[] n_par = new int[par.length];
	double[] n_score = new double[par.length];
	for(int i = 0; i < par.length; i++) {
	    result[i] = -1;
	    n_par[i] = -1;
	    n_score[i] = Double.NEGATIVE_INFINITY;
	}

	boolean[][] isChild = calcChilds(par);

	for(int i = 1; i < n_par.length; i++) {
	    double max = Double.NEGATIVE_INFINITY;
	    int wh = -1;
	    for(int j = 0; j < n_par.length; j++) {
		if(i == j || par[i] == j || isChild[i][j]) continue;
		if(scoreMatrix[j][i] > max) { max = scoreMatrix[j][i]; wh = j; }
	    }
	    n_par[i] = wh;
	    n_score[i] = max;
	}

	for(int k = 0; k < K; k++) {
	    double max = Double.NEGATIVE_INFINITY;
	    int wh = -1;
	    int whI = -1;
	    for(int i = 0; i < n_par.length; i++) {
		if(n_par[i] == -1) continue;
		double score = scoreMatrix[n_par[i]][i];
		if(score > max) {
		    max = score; whI = i; wh = n_par[i];
		}
	    }

	    if(max == Double.NEGATIVE_INFINITY)
		break;
	    result[whI] = wh;
	    n_par[whI] = -1;
	}

	return result;
    }

    private static TIntIntHashMap chuLiuEdmonds(double[][] scoreMatrix, boolean[] curr_nodes, 
						int[][] oldI, int[][] oldO, boolean print,
						TIntIntHashMap final_edges, TIntIntHashMap[] reps) {
		
	// need to construct for each node list of nodes they represent (here only!)
		
	int[] par = new int[curr_nodes.length];
	int numWords = curr_nodes.length;
		
	// create best graph
	par[0] = -1;
	for(int i = 1; i < par.length; i++) {
	    // only interested in current nodes
	    if(!curr_nodes[i]) continue;
	    double maxScore = scoreMatrix[0][i];
	    par[i] = 0;
	    for(int j = 0; j < par.length; j++) {
		if(j == i) continue;
		if(!curr_nodes[j]) continue;
		double newScore = scoreMatrix[j][i];
		if(newScore > maxScore) {
		    maxScore = newScore;
		    par[i] = j;
		}
	    }
	}
		
	if(print) {
	    System.out.println("After init");
	    for(int i = 0; i < par.length; i++) {
		if(curr_nodes[i])
		    System.out.print(par[i] + "|" + i + " ");
	    }
	    System.out.println();
	}
		
	//Find a cycle
	ArrayList cycles = new ArrayList();
	boolean[] added = new boolean[numWords];
	for(int i = 0; i < numWords && cycles.size() == 0; i++) {
	    // if I have already considered this or
	    // This is not a valid node (i.e. has been contracted)
	    if(added[i] || !curr_nodes[i]) continue;
	    added[i] = true;
	    TIntIntHashMap cycle = new TIntIntHashMap();
	    cycle.put(i,0);
	    int l = i;
	    while(true) {
		if(par[l] == -1) {
		    added[l] = true;
		    break;
		}
		if(cycle.contains(par[l])) {
		    cycle = new TIntIntHashMap();
		    int lorg = par[l];
		    cycle.put(lorg,par[lorg]);
		    added[lorg] = true;
		    int l1 = par[lorg];
		    while(l1 != lorg) {
			cycle.put(l1,par[l1]);
			added[l1] = true;
			l1 = par[l1];
						
		    }
		    cycles.add(cycle);
		    break;
		}
		cycle.put(l,0);
		l = par[l];
		if(added[l] && !cycle.contains(l))
		    break;
		added[l] = true;
	    }
	}
		
	// get all edges and return them
	if(cycles.size() == 0) {
	    //System.out.println("TREE:");
	    for(int i = 0; i < par.length; i++) {
		if(!curr_nodes[i]) continue;
		if(par[i] != -1) {
		    int pr = oldI[par[i]][i];
		    int ch = oldO[par[i]][i];
		    final_edges.put(ch,pr);
		    //System.out.print(pr+"|"+ch + " ");
		}
		else
		    final_edges.put(0,-1);
	    }
	    //System.out.println();
	    return final_edges;
	}

	int max_cyc = 0;
	int wh_cyc = 0;
	for(int i = 0; i < cycles.size(); i++) {
	    TIntIntHashMap cycle = (TIntIntHashMap)cycles.get(i);
	    if(cycle.size() > max_cyc) { max_cyc = cycle.size(); wh_cyc = i; }
	}
		
	TIntIntHashMap cycle = (TIntIntHashMap)cycles.get(wh_cyc);
	int[] cyc_nodes = cycle.keys();
	int rep = cyc_nodes[0];
		
	if(print) {
	    System.out.println("Found Cycle");
	    for(int i = 0; i < cyc_nodes.length; i++)
		System.out.print(cyc_nodes[i] + " ");
	    System.out.println();
	}

	double cyc_weight = 0.0;
	for(int j = 0; j < cyc_nodes.length; j++) {
	    cyc_weight += scoreMatrix[par[cyc_nodes[j]]][cyc_nodes[j]];
	}
		
		
	for(int i = 0; i < numWords; i++) {
			
	    if(!curr_nodes[i] || cycle.contains(i)) continue;
			
			
	    double max1 = Double.NEGATIVE_INFINITY;
	    int wh1 = -1;
	    double max2 = Double.NEGATIVE_INFINITY;
	    int wh2 = -1;
			
	    for(int j = 0; j < cyc_nodes.length; j++) {
		int j1 = cyc_nodes[j];
				
		if(scoreMatrix[j1][i] > max1) {
		    max1 = scoreMatrix[j1][i];
		    wh1 = j1;//oldI[j1][i];
		}
				
		// cycle weight + new edge - removal of old
		double scr = cyc_weight + scoreMatrix[i][j1] - scoreMatrix[par[j1]][j1];
		if(scr > max2) {
		    max2 = scr;
		    wh2 = j1;//oldO[i][j1];
		}
	    }
			
	    scoreMatrix[rep][i] = max1;
	    oldI[rep][i] = oldI[wh1][i];//wh1;
	    oldO[rep][i] = oldO[wh1][i];//oldO[wh1][i];
	    scoreMatrix[i][rep] = max2;
	    oldO[i][rep] = oldO[i][wh2];//wh2;
	    oldI[i][rep] = oldI[i][wh2];//oldI[i][wh2];
			
	}
		
	TIntIntHashMap[] rep_cons = new TIntIntHashMap[cyc_nodes.length];
	for(int i = 0; i < cyc_nodes.length; i++) {
	    rep_cons[i] = new TIntIntHashMap();
	    int[] keys = reps[cyc_nodes[i]].keys();
	    Arrays.sort(keys);
	    if(print) System.out.print(cyc_nodes[i] + ": ");
	    for(int j = 0; j < keys.length; j++) {
		rep_cons[i].put(keys[j],0);
		if(print) System.out.print(keys[j] + " ");
	    }
	    if(print) System.out.println();
	}
		
	// don't consider not representative nodes
	// these nodes have been folded
	for(int i = 1; i < cyc_nodes.length; i++) {
	    curr_nodes[cyc_nodes[i]] = false;
	    int[] keys = reps[cyc_nodes[i]].keys();
	    for(int j = 0; j < keys.length; j++)
		reps[rep].put(keys[j],0);
	}
		
	chuLiuEdmonds(scoreMatrix,curr_nodes,oldI,oldO,print,final_edges,reps);
		
	// check each node in cycle, if one of its representatives
	// is a key in the final_edges, it is the one.
	int wh = -1;
	boolean found = false;
	for(int i = 0; i < rep_cons.length && !found; i++) {
	    int[] keys = rep_cons[i].keys();
	    for(int j = 0; j < keys.length && !found; j++) {
		if(final_edges.contains(keys[j])) {
		    wh = cyc_nodes[i];
		    found = true;
		}
	    }
	}
		
	int l = par[wh];
	while(l != wh) {
	    int ch = oldO[par[l]][l];
	    int pr = oldI[par[l]][l];
	    final_edges.put(ch,pr);
	    l = par[l];
	}
		
	if(print) {
	    int[] keys = final_edges.keys();
	    Arrays.sort(keys);
	    for(int i = 0; i < keys.length; i++)
		System.out.print(final_edges.get(keys[i])+"|"+keys[i]+" ");
	    System.out.println();
	}
		
	return final_edges;
		
    }
	
    

}

File: src/main/scala/mstparser/DependencyDecoder.scala
package mstparser

import scala.collection.mutable.Buffer
import com.google.common.collect.MinMaxPriorityQueue

class DependencyDecoder(protected val pipe: DependencyPipe) {
  protected def getTypes(probsNt: Array[Array[Array[Array[Double]]]], len: Int) =
    Array.tabulate(len, len) {
      case (i, j) if i == j => 0
      case (i, j) => probsNt(i).zip(probsNt(j)).zipWithIndex.maxBy {
        case ((pi, pj), _) => if (i < j) pi(0)(1) + pj(0)(0) else pi(1)(1) + pj(1)(0)
      }._2
    }

  protected def calcChilds(parse: Seq[Int]): Seq[Seq[Boolean]] = {
    val isChild = IndexedSeq.fill(parse.size)(Buffer.fill(parse.size)(false))

    parse.zipWithIndex.drop(1).foreach { case (v, i) =>
      var l = v
      while (l != -1) {
        isChild(l)(i) = true
        l = parse(l)
      }
    }
    isChild
  }

  // static type for each edge: run time O(n^3 + Tn^2) T is number of types
  def decodeProjective(
    len: Int,
    fvs: Array[Array[Array[FeatureVector]]],
    probs: Array[Array[Array[Double]]],
    fvsNt: Array[Array[Array[Array[FeatureVector]]]],
    probsNt: Array[Array[Array[Array[Double]]]],
    kBest: Int
  ): Seq[(FeatureVector, (IndexedSeq[Int], IndexedSeq[Int]))] = {
    val staticTypes = if (this.pipe.labeled) Some(this.getTypes(probsNt, len)) else None
    val pf = new KBestParseForest(len - 1, kBest)
    val qb = MinMaxPriorityQueue.orderedBy(Ordering.by[ParseForestItem, Double](_.prob).reverse).maximumSize(kBest)
    val q0 = qb.create[ParseForestItem]
    val q1 = qb.create[ParseForestItem]

    (1 until len).foreach { j =>
      (0 until len - j).foreach { s =>
        val t = s + j
        val (type1, type2) = staticTypes.map(ts => (ts(s)(t), ts(t)(s))).getOrElse((0, 0))

        (s to t).foreach { r =>
          if (r != t) {
            val bs = pf.complete(s)(r)
            val cs = pf.complete(t)(r + 1)

            pf.bestPairs(bs, cs)(_.prob).take(kBest).foreach { case ((b, c), v) =>

              var finProb = v + probs(s)(t)(0)
              var finFv = fvs(s)(t)(0)
              if (this.pipe.labeled) {
                finFv = fvsNt(s)(type1)(0)(1).cat(fvsNt(t)(type1)(0)(0).cat(finFv))
                finProb += probsNt(s)(type1)(0)(1) + probsNt(t)(type1)(0)(0)
              }
              q0.add(ArcItem(s, t, type1, finProb, finFv, b, c)) 

              finProb = v + probs(s)(t)(1)
              finFv = fvs(s)(t)(1)
              if (this.pipe.labeled) {
                finFv = fvsNt(t)(type2)(1)(1).cat(fvsNt(s)(type2)(1)(0).cat(finFv))
                finProb += probsNt(t)(type2)(1)(1) + probsNt(s)(type2)(1)(0)
              }
              q1.add(ArcItem(t, s, type2, finProb, finFv, b, c)) 
            }
          }
        }
        pf.incomplete(s)(t) = IndexedSeq.fill(q0.size)(q0.pollFirst)
        pf.incomplete(t)(s) = IndexedSeq.fill(q1.size)(q1.pollFirst)

        (s to t).foreach { r =>
          if (r != s) {
            val bs = pf.incomplete(s)(r)
            val cs = pf.complete(r)(t)

            pf.bestPairs(bs, cs)(_.prob).take(kBest).foreach { case ((b, c), _) =>
              q0.add(CompleteItem(b, c))
            }
          }

          if (r != t) {
            val bs = pf.complete(r)(s)
            val cs = pf.incomplete(t)(r)

            pf.bestPairs(bs, cs)(_.prob).take(kBest).foreach { case ((b, c), _) =>
              q1.add(CompleteItem(b, c))
            }
          }
        }
        pf.complete(s)(t) = IndexedSeq.fill(q0.size)(q0.pollFirst)
        pf.complete(t)(s) = IndexedSeq.fill(q1.size)(q1.pollFirst)
      }
    }
    pf.getBestParses
  }

  protected def getKChanges(parse: Array[Int], scores: Array[Array[Double]], k: Int) = {
    val result = Array.fill(parse.size)(-1)
    val isChild = this.calcChilds(parse)

    val nParse = (-1 +: parse.zipWithIndex.tail.map { case (p, i) =>
      val valid = (0 until parse.size).filter(j =>
        i != j && p != j && !isChild(i)(j)
      )
      if (valid.isEmpty) -1 else valid.maxBy(j => scores(j)(i))
    }).toBuffer

    (0 until k).foreach { i =>
      val pairs = nParse.zipWithIndex.filter(_._1 > -1)
      if (!pairs.isEmpty) {
        val (mp, mi) = pairs.maxBy { case (np, ni) =>
          scores(np)(ni)
        }

        result(mi) = mp
        nParse(mi) = -1
      }
    }

    result
  }

  def decodeNonProjective(
    len: Int,
    fvs: Array[Array[Array[FeatureVector]]],
    probs: Array[Array[Array[Double]]],
    fvsNt: Array[Array[Array[Array[FeatureVector]]]],
    probsNt: Array[Array[Array[Array[Double]]]],
    kBest: Int
  ): Seq[(FeatureVector, (IndexedSeq[Int], IndexedSeq[Int]))] = {
    val staticTypes = if (this.pipe.labeled) Some(this.getTypes(probsNt, len)) else None

    val scores = Array.tabulate(len, len) {
      case (i, j) if i < j =>
        probs(i)(j)(0) + staticTypes.map(ts =>
          probsNt(i)(ts(i)(j))(0)(1) + probsNt(j)(ts(i)(j))(0)(0)
        ).getOrElse(0.0)

      case (i, j) =>
        probs(j)(i)(1) + staticTypes.map(ts =>
          probsNt(i)(ts(i)(j))(1)(1) + probsNt(j)(ts(i)(j))(1)(0)
        ).getOrElse(0.0)
    }

    val parse = Array.ofDim[Int](len)
    this.chuLiuEdmonds(scores).foreach { case (i, v) => parse(i) = v }

    val parFin = parse +: this.getKChanges(
      parse, scores, math.min(kBest, parse.length)
    ).zipWithIndex.filter(_._1 > -1).map {
      case (p, i) => parse.updated(i, p)
    }

    val fvFin = parFin.map { p =>
      p.zipWithIndex.map {
        case (-1, _) => new FeatureVector
        case (j, i)  =>
          val f = fvs(if (i < j) i else j)(if (i < j) j else i)(if (i < j) 1 else 0)
          staticTypes.map(ts =>
            f.cat(fvsNt(i)(ts(j)(i))(if (i < j) 1 else 0)(0)).cat(
                  fvsNt(j)(ts(j)(i))(if (i < j) 1 else 0)(1))
          ).getOrElse(f)
      }
    }

    //val fin = Array.fill(parFin.size)(new FeatureVector)
    //val result = Array.fill(parFin.size)("")

    fvFin.zip(parFin).map { case (fvs, hs) =>
      val (f, (p, l)) = fvs.zip(hs).zipWithIndex.drop(1).foldLeft(
        new FeatureVector, (Array.fill(-1)(hs.size), Array.fill(-1)(hs.size)) 
      ) { case ((f, (parse, labels)), ((fv, h), i)) => {
          parse(i) = h
          labels(i) = staticTypes.map(_(h)(i)).getOrElse(0) 
          (fv.cat(f), (parse, labels))
        //fin(i) = fv.cat(fin(i))
        //result(i) += "%d|%d:%d ".format(parse, j, staticTypes.map(_(parse)(j)).getOrElse(0))
        }
      }
      (f, (wrapIntArray(p), wrapIntArray(l)))
    }

    //fin.zip(result.map(_.trim))
  }

  def chuLiuEdmonds(scores: Array[Array[Double]]) = {
    def cLE(
      scores: Array[Array[Double]],
      oldI: Array[Array[Int]],
      oldO: Array[Array[Int]],
      nodes: Set[Int],
      edges: Map[Int, Int],
      reps: IndexedSeq[Set[Int]],
      printing: Boolean
    ): Map[Int, Int] = {
      // Need to construct for each node list of nodes they represent (here only!)
      val len = scores.size

      val is = (0 until len)
      val as = is.filter(nodes)

      val parse = -1 +: is.tail.map {
        case i if !nodes(i) => 0
        case i => as.filter(_ != i).maxBy(scores(_)(i))
      }

      val cycles = Buffer.empty[Set[Int]]
      val added = collection.mutable.Set.empty[Int]

      var i = 0
      while (i < len && cycles.isEmpty) {
        if (!added(i) && nodes(i)) {
          added += i
          val cycle = collection.mutable.Set(i)

          var j = i
          var found = false
          while (!found) {
            if (parse(j) == -1) {
              added += j
              found = true
            } else {
              if (cycle(parse(j))) {
                cycle.clear()
                val k = parse(j)
                cycle += k
                added += k

                var l = parse(k)
                while (l != k) {
                  cycle += l
                  added += l
                  l = parse(l)
                }

                cycles += cycle.toSet
                found = true
              } else {
                cycle += j
                j = parse(j)
                if (added(j) && !cycle(j))
                  found = true
                else
                  added(j) = true
              }
            }
          }
        }
        i += 1
      }

      if (cycles.isEmpty) {
        edges ++
        parse.zipWithIndex.filter(pi => nodes(pi._2)).map {
          case (-1, _) => 0 -> -1
          case (p, i)  => oldO(p)(i) -> oldI(p)(i)
        }.toMap
      }
      else {
        val biggest = cycles.maxBy(_.size)
        val cNodes = biggest.toSeq.sorted.reverse
        val weight = cNodes.map(i => scores(parse(i))(i)).sum

        as.filterNot(biggest.contains).foreach { i =>
          val (ai, aw) = cNodes.map(j => j -> scores(j)(i)).maxBy(_._2)
          val (bi, bw) = cNodes.map(j => j -> (weight + scores(i)(j) - scores(parse(j))(j))).maxBy(_._2)

          scores(cNodes.head)(i) = aw
          oldI(cNodes.head)(i) = oldI(ai)(i)
          oldO(cNodes.head)(i) = oldO(ai)(i)

          scores(i)(cNodes.head) = bw
          oldI(i)(cNodes.head) = oldI(i)(bi)
          oldO(i)(cNodes.head) = oldO(i)(bi)
        }

        val nRep = cNodes.tail.map(reps).foldLeft(reps(cNodes.head))(_ ++ _)

        val nEdges = cLE(
          scores, oldI, oldO,
          nodes -- cNodes.tail,
          edges,
          reps.updated(cNodes.head, nRep),
          printing
        )

        // TODO: Could be more efficient, since we don't need the whole
        // intersection.
        val n = cNodes.find(i => !reps(i).intersect(nEdges.keySet).isEmpty).getOrElse(-1)

        nEdges ++ Stream.iterate(parse(n))(parse(_)).takeWhile(_ != n).map(i =>
          oldO(parse(i))(i) -> oldI(parse(i))(i)
        ).toMap
      }
    }

    val len = scores.size

    cLE(
      scores.map(_.clone),
      Array.tabulate(len, len)((i, _) => i),
      Array.tabulate(len, len)((_, j) => j),
      (0 until len).toSet,
      Map.empty[Int, Int],
      IndexedSeq.tabulate(len)(Set(_)),
      false
    )
  }
}

